control-plane-operator/hostedclusterconfigoperator/controllers/resources/kas/admissionpolicies.go

```diff
@@ -0,0 +1,118 @@
+
+	if err := reconcileConfigValidatingAdmissionPolicy(ctx, hcp, client, createOrUpdate); err != nil {
+		return fmt.Errorf("failed to reconcile Config Validating Admission Policy: %v", err)
+	}
+
+	if err := reconcileICSPValidatingAdmissionPolicy(ctx, client, createOrUpdate); err != nil {
+		return fmt.Errorf("failed to reconcile ICSP Validating Admission Policy: %v", err)
+	}
+
+	return nil
+}
+
+func reconcileICSPValidatingAdmissionPolicy(ctx context.Context, client client.Client, createOrUpdate upsert.CreateOrUpdateFN) error {
+	icspAdmissionPolicy := AdmissionPolicy{Name: AdmissionPolicyNameICSP}
+	icspApiVersion := []string{"v1alpha1"}
+	icspApiGroup := []string{"operator.openshift.io"}
+	icspResources := []string{"imagecontentsourcepolicies"}
+
+	icspAdmissionPolicy.Validations = []k8sadmissionv1beta1.Validation{HCCOUserValidation}
+	icspAdmissionPolicy.MatchConstraints = manifests.VAPolicyMatchConstraints(icspResources, icspApiVersion, icspApiGroup, defaultAdmissionPoliciesOperations)
+	if err := icspAdmissionPolicy.reconcileAdmissionPolicy(ctx, client, createOrUpdate); err != nil {
+		return fmt.Errorf("error reconciling ICSP Validating Admission Policy: %v", err)
+	}
+
+	return nil
+}
+
+func reconcileConfigValidatingAdmissionPolicy(ctx context.Context, hcp *hyperv1.HostedControlPlane, client client.Client, createOrUpdate upsert.CreateOrUpdateFN) error {
+	// Config AdmissionPolicy
+	configAdmissionPolicy := AdmissionPolicy{Name: AdmissionPolicyNameConfig}
+	configApiVersion := []string{"v1"}
+	configApiGroup := []string{"config.openshift.io"}
+	configResources := []string{
+		"apiservers",
+		"authentications",
+		"featuregates",
+		"images",
+		"imagedigestmirrorsets",
+		"imagecontentpolicies",
+		"infrastructures",
+		"ingresses",
+		"proxies",
+		"schedulers",
+		"networks",
+		"oauths",
+	}
+
+	if hcp.Spec.OLMCatalogPlacement == hyperv1.ManagementOLMCatalogPlacement {
+		configResources = append(configResources, "operatorhubs")
+	}
+	configAdmissionPolicy.Validations = []k8sadmissionv1beta1.Validation{HCCOUserValidation}
+	configAdmissionPolicy.MatchConstraints = manifests.VAPolicyMatchConstraints(configResources, configApiVersion, configApiGroup, defaultAdmissionPoliciesOperations)
+	if err := configAdmissionPolicy.reconcileAdmissionPolicy(ctx, client, createOrUpdate); err != nil {
+		return fmt.Errorf("error reconciling Config Validating Admission Policy: %v", err)
+	}
+
+	return nil
+}
+
+func (ap *AdmissionPolicy) reconcileAdmissionPolicy(ctx context.Context, client client.Client, createOrUpdate upsert.CreateOrUpdateFN) error {
+	vap := manifests.ValidatingAdmissionPolicy(ap.Name)
+	vap.Spec = k8sadmissionv1beta1.ValidatingAdmissionPolicySpec{
+		MatchConstraints: ap.MatchConstraints,
+		Validations:      ap.Validations,
+	}
+
+	if _, err := createOrUpdate(ctx, client, vap, func() error { return nil }); err != nil {
+		return fmt.Errorf("failed to create/update Validating Admission Policy with name %s: %v", ap.Name, err)
+	}
```

Comment:
changes to the object should be done inside the mutate functions (last argument of `createOrUpdate`). Otherwise, out of band changes will not be kept and not reconciled over. 

```suggestion
	if _, err := createOrUpdate(ctx, client, vap, func() error { 
	  // update spec here
	  vap.Spec = ....
	  return nil 
	}); err != nil {
		return fmt.Errorf("failed to create/update Validating Admission Policy with name %s: %v", ap.Name, err)
	}
```